# Exploring Ebay Car Sales

**This Jupyter notebook project focuses on analyzing a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.**

**The goal of this project is to clean and prepare the dataset for analysis, and then use exploratory data analysis techniques to gain insights into the used car market in Germany.**

**By examining the dataset, we hope to uncover interesting patterns and trends related to the prices, mileage, and condition of used cars, as well as the most popular makes and models.**


In [1]:
# import libraries that are needed 
import pandas as pd
import numpy as np

In [2]:
# read in the file
path = '/DataQuest_Projects/data_files/'
autos = pd.read_csv(path + 'autos.csv', encoding= 'Windows-1252')

In [3]:
# show an overview of the DataFrame's structure and contents 
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

In [4]:
autos.head()

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

**From the above dataframe summary, we can see that the dataset contains `50,000 entries` with `20 columns`.** 

**Some columns have missing values, such as `"vehicleType"`, `"gearbox"`, `"model"`, `"fuelType"`, and `"notRepairedDamage"`.**

**Additionally, some columns have non-numeric data types, such as `"price"` and `"odometer"`, which are currently stored as objects instead of numerical values.** 

**We will need to clean and preprocess the data in order to make it suitable for analysis. Furthermore, we can see that the dataset includes information on various attributes of used cars, such as their make, model, year of registration, power, fuel type, and odometer readings.**

In [5]:
# show the columns names
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [6]:
# renaming columns
autos.rename(columns={"yearOfRegistration":"registration_year", "monthOfRegistration":"registration_month", "notRepairedDamage":"unrepaired_damage", "dateCreated":"ad_created"}, inplace=True)


In [7]:
# convert columns from camel case to snake case using the module pycase

snake_case_list = []

for column_name in autos.columns:
    new_str = ""
    for char in column_name: 
        if char.isupper():
            change_char = '_' + char.swapcase()
            new_str += change_char
        else:
            new_str += char
    snake_case_list.append(new_str)  

autos.columns = snake_case_list

In [8]:
# show the first few lines of the Dataframe
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_p_s   model   odometer  registration_month fuel_type  \
0    manuell        158  andere  150,000km                   3       lpg   
1  automatik        286     7er  150,000km                   6    benzin   
2    manuell        102    golf   70,000km                   7    benzin   
3  automatik         71  fortwo   70,000km                   6    benzin   
4    manuell          0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

**The columns are renamed to more descriptive names, and the format is changed from camel case to snake case using the Python library pycase.** 

**Renaming the columns is useful for making the DataFrame more readable and for removing any ambiguities in column names. Converting the names to snake case is a common practice in Python programming, as it makes the column names easier to read and write.** 

**By using the pycase library, the process of converting the column names from camel case to snake case is automated, saving time and reducing the chance of errors.** 

**The changes are made in place using the `'inplace=True'` parameter in the `rename()` function, which modifies the DataFrame directly.**

In [9]:
# show desciptive statistics, use 'all' to view categorical and numerical columns 
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 11:37:04  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox     power_p_s  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

## Based on the above statistics here are some observations:

    The "seller" and "offer_type" columns have mostly one value (49999 out of 50000) and could be candidates to be dropped.
    The "name" column may need further investigation to determine if it contains any useful information.
    The "price" and "odometer" columns contain numeric data stored as text (with the units "USD" and "km" included) and may need to be cleaned by removing the non-numeric characters and converting the data type to float or int.
    The "registration_year" column has a maximum value of 9999, which seems unrealistic and may need to be investigated and cleaned.

In [10]:
# clean price column and convert to float dtype 

clean_price_list = []

for price in autos['price']:
    clean_price = price.replace("$", "").replace(",", "")
    clean_price_list.append(int(clean_price))

autos["price"] = clean_price_list    

In [11]:
# clean odometer column and convert to float

clean_odometer_list = []

for num in autos["odometer"]:
    clean_num = num.replace('km', '').replace(',', '')
    clean_odometer_list.append(float(clean_num))
    
autos["odometer"] = clean_odometer_list    

In [12]:
# rename price and odometer column

autos.rename(columns={'price':'price_usd', 'odometer':'odometer_km'}, inplace=True)

## Exploring and Analyzing `price_usd` and `odometer_km` columns 

```
for each column we will do the following:
    -check the unique values 
    -view the statistics using `describe` method
    -use `value_counts` and remove any outliers if needed
```



In [13]:
# odometer column 
# view the unique values
print('Odometer column has', autos['odometer_km'].unique().shape[0], 'unique values')
print('\nUnique Values\n')
print(autos['odometer_km'].unique())
autos['odometer_km'].describe()

Odometer column has 13 unique values

Unique Values

[150000.  70000.  50000.  80000.  10000.  30000. 125000.  90000.  20000.
  60000.   5000. 100000.  40000.]


count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

## Based on the summary statistics provided for the column `"odometer_km"`, I can make the following observations:

**The dataset has a total of `50,000` values for the "odometer_km" column.
The most common odometer reading in the dataset is `150,000` km, which appears in over `30,000` entries.
The mean odometer reading is approximately `125,732` km, indicating that the average vehicle in the dataset has been driven for a considerable distance.
The standard deviation is `40,042` km, which is relatively high, indicating that the odometer readings are spread out over a wide range.**
    
**Overall, the dataset appears to have a wide range of odometer readings, but most of the vehicles have relatively high mileage, with a median value of 150,000 km.**

In [14]:
# price column
# view the unique values
print('Price column has', autos['price_usd'].unique().shape[0], 'unique values')
print('')
autos['price_usd'].describe()


Price column has 2357 unique values



count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_usd, dtype: float64

In [15]:
# remove any outliers that are less then 500 usd and are greater then 50,000 usd
autos = autos[autos['price_usd'].between(500,50001, inclusive='neither')]

In [16]:
# view the statistics 
autos['price_usd'].describe()

count    44130.000000
mean      6078.689259
std       6821.204820
min        501.000000
25%       1550.000000
50%       3500.000000
75%       7950.000000
max      50000.000000
Name: price_usd, dtype: float64

## Based on the summary statistics provided for the column `"price_usd"`, I can make the following observations:

**The dataset has a total of `44,130` values for the `"price_usd"` column.
The mean price is approximately `6,078.69` USD.
The standard deviation is relatively high at `6,821.20` USD, indicating that the prices are spread out over a wide range.
The minimum price is `501` USD, which indicates that there are some relatively inexpensive items in the dataset.
The 25th percentile value is `1,550` USD, which means that `25%` of the items in the dataset are priced at or below this value.
The median price is `3,500` USD, indicating that half of the items in the dataset are priced at or below this value.
The 75th percentile value is `7,950` USD, indicating that `75%` of the items in the dataset are priced at or below this value.
The maximum price is `50,000` USD, indicating that there are some relatively expensive items in the dataset.**

**Overall, the dataset appears to have a wide range of prices, with a relatively high standard deviation. However, most of the values are clustered around the lower end of the price range, as indicated by the relatively low median price.**

In [17]:
# select only the date and then calculate the distribution of values as percentages

def get_value_counts_by_date(column_name):
    return autos[column_name].str[:10].value_counts(normalize=True, dropna=False).sort_index()


In [18]:
# distribution of 'date_crawled' column
print(get_value_counts_by_date('date_crawled'))

2016-03-05    0.025357
2016-03-06    0.014140
2016-03-07    0.036030
2016-03-08    0.032880
2016-03-09    0.032880
2016-03-10    0.032857
2016-03-11    0.032925
2016-03-12    0.037480
2016-03-13    0.015681
2016-03-14    0.036302
2016-03-15    0.034036
2016-03-16    0.029277
2016-03-17    0.031135
2016-03-18    0.012848
2016-03-19    0.034965
2016-03-20    0.038137
2016-03-21    0.037752
2016-03-22    0.032993
2016-03-23    0.032246
2016-03-24    0.028983
2016-03-25    0.031385
2016-03-26    0.032744
2016-03-27    0.030954
2016-03-28    0.034829
2016-03-29    0.033333
2016-03-30    0.033197
2016-03-31    0.031634
2016-04-01    0.034013
2016-04-02    0.035849
2016-04-03    0.038862
2016-04-04    0.036574
2016-04-05    0.013166
2016-04-06    0.003172
2016-04-07    0.001382
Name: date_crawled, dtype: float64


**The following observations can be made for the frequency distribution of the `"date_crawled"` column**

**The date range covered in the dataset is from `March 5, 2016, to April 7, 2016`.
The `highest proportion of rows (3.89%) is on April 3, 2016`, while the `lowest proportion (0.14%) is on April 7, 2016`.
The proportion of rows tends to be higher during the weekends (March 5-6, March 12-13, March 19-20, and April 2-3) compared to weekdays.
There is a gradual increase in the proportion of rows until April 3, 2016, after which there is a gradual decrease until April 7, 2016.**


In [19]:
# Set the max_rows option to None to display the full dataframe
pd.set_option('display.max_rows', None)

# distribution of 'date_crawled' column
print(get_value_counts_by_date('ad_created'))

2015-06-11    0.000023
2015-08-10    0.000023
2015-09-09    0.000023
2015-11-10    0.000023
2015-12-05    0.000023
2015-12-30    0.000023
2016-01-03    0.000023
2016-01-07    0.000023
2016-01-10    0.000045
2016-01-13    0.000023
2016-01-14    0.000023
2016-01-16    0.000023
2016-01-22    0.000023
2016-01-27    0.000068
2016-01-29    0.000023
2016-02-01    0.000023
2016-02-02    0.000045
2016-02-05    0.000045
2016-02-07    0.000023
2016-02-08    0.000023
2016-02-09    0.000023
2016-02-11    0.000023
2016-02-12    0.000045
2016-02-14    0.000045
2016-02-16    0.000023
2016-02-17    0.000023
2016-02-18    0.000045
2016-02-19    0.000068
2016-02-20    0.000045
2016-02-21    0.000045
2016-02-22    0.000023
2016-02-23    0.000091
2016-02-24    0.000045
2016-02-25    0.000068
2016-02-26    0.000045
2016-02-27    0.000136
2016-02-28    0.000227
2016-02-29    0.000181
2016-03-01    0.000091
2016-03-02    0.000113
2016-03-03    0.000861
2016-03-04    0.001473
2016-03-05    0.022910
2016-03-06 

**The following observations can be made for the frequency distribution of the `"ad_created"` column. 
The index of the series represents the dates on which the ads were created, 
while the values represent the proportion of ads created on that date in the dataset.**

**The proportion of ads created seems to increase gradually over time until a peak around early March, after which the proportion decreases.
There are several spikes in the data, where the proportion of ads created on a particular date is significantly higher than the surrounding dates. For example, there is a noticeable spike on February 28th, where the proportion of ads created is almost 10 times higher than the surrounding dates. This may indicate a specific event or promotion that led to a surge in ad creation or a potential spambot was used to create ads.
The proportion of ads created on weekends (Saturday and Sunday) appears to be higher than on weekdays, suggesting that weekends may be a more popular time for ad creation.**


In [20]:
# distribution of 'date_crawled' column
print(get_value_counts_by_date('last_seen'))

2016-03-05    0.001065
2016-03-06    0.004056
2016-03-07    0.004917
2016-03-08    0.006979
2016-03-09    0.009427
2016-03-10    0.010265
2016-03-11    0.011965
2016-03-12    0.023567
2016-03-13    0.008883
2016-03-14    0.012350
2016-03-15    0.015522
2016-03-16    0.016157
2016-03-17    0.027646
2016-03-18    0.007410
2016-03-19    0.015386
2016-03-20    0.020236
2016-03-21    0.020780
2016-03-22    0.021255
2016-03-23    0.018355
2016-03-24    0.019533
2016-03-25    0.018513
2016-03-26    0.016361
2016-03-27    0.015318
2016-03-28    0.020508
2016-03-29    0.021255
2016-03-30    0.024065
2016-03-31    0.023453
2016-04-01    0.023046
2016-04-02    0.025108
2016-04-03    0.024700
2016-04-04    0.024428
2016-04-05    0.126762
2016-04-06    0.226331
2016-04-07    0.134398
Name: last_seen, dtype: float64


**The following observations can be made for the frequency distribution of the `"last_seen"` column.**

**Upon inspection of the data, we can see that the percentage of car listings that were last seen remained relatively stable from `March 5th to April 4th, 2016`**

**However, there is a significant increase in the percentage of car listings that were last seen on `April 5th, 2016`, with a value of 0.126762. This value is much higher than the values observed in the previous days. The percentage remains high on `April 6th, 2016`, with a value of 0.226331. The value then decreases slightly on `April 7th, 2016`, with a value of 0.134398.**

**This sudden increase in the percentage of car listings that were last seen on April 5th and April 6th, 2016, may indicate a surge in activity on the platform, which could be due to various reasons such as promotions, events or changes to the website's algorithm.**

In [21]:
# distribution of 'registration_year' column
autos['registration_year'].describe()

count    44130.000000
mean      2005.111262
std         89.633152
min       1000.000000
25%       2000.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

**The following observations can be made for the frequency distribution of the `"registration_year"` column.**

**The `mean value of 2005.11` indicates that on average, the cars were registered around the year 2005. 
The standard deviation of 89.63 tells us that the data is relatively spread out around the mean, which could suggest that there may be a significant number of older and newer vehicles in the dataset.**

**The `minimum value of 1000` and `maximum value of 9999` appear to be outliers, as it is not possible for a car to have been registered before the invention of automobiles or in the far future. This may indicate that there are some erroneous data points in the dataset and further investigation is needed.**

**The 25th, 50th, and 75th percentiles (also known as the first, second, and third quartiles) of 2000, 2004, and 2008 respectively suggest that the majority of the cars in the dataset were registered between the years 2000 and 2008.**

In [22]:
# Select the rows where registration year is less than 1885 or greater than 2016,
# and count the occurrences of each value in the resulting subset
# This allows us to see the distribution of invalid registration years in the dataset

autos[(autos['registration_year']<1885) | (autos['registration_year']>2016)]['registration_year'].value_counts().sort_index()

1000       1
1001       1
2017    1275
2018     458
2019       1
2800       1
4100       1
4500       1
5000       2
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

**The highest acceptable value would be the 2016, as the data was collected between March 2016 - April 2016. Therefore, any value greater than 2016 seems unrealistic and should be removed from the dataset. 
On the other hand, the lowest acceptable value for the `registration_year` column would be the year in which the first automobile was invented, which was 1885. Therefore, any value less than 1885 is clearly unrealistic and should also be removed from the dataset.**

In [23]:
# remove the rows outside the bounds of 1885 - 2016
autos = autos[(autos['registration_year']>1885) & (autos['registration_year']<=2016)]

In [26]:
autos['registration_year'].value_counts(normalize=True).sort_index()

1910    0.000024
1927    0.000024
1929    0.000024
1931    0.000024
1934    0.000047
1937    0.000094
1938    0.000024
1939    0.000024
1941    0.000047
1943    0.000024
1948    0.000024
1950    0.000024
1951    0.000024
1952    0.000024
1953    0.000024
1954    0.000047
1956    0.000094
1957    0.000047
1958    0.000071
1959    0.000142
1960    0.000425
1961    0.000142
1962    0.000094
1963    0.000165
1964    0.000212
1965    0.000401
1966    0.000472
1967    0.000566
1968    0.000590
1969    0.000401
1970    0.000779
1971    0.000543
1972    0.000731
1973    0.000543
1974    0.000566
1975    0.000425
1976    0.000495
1977    0.000495
1978    0.000991
1979    0.000802
1980    0.001746
1981    0.000637
1982    0.000967
1983    0.001156
1984    0.001156
1985    0.001982
1986    0.001534
1987    0.001557
1988    0.002949
1989    0.003563
1990    0.006064
1991    0.006465
1992    0.006913
1993    0.007550
1994    0.010759
1995    0.018546
1996    0.023619
1997    0.033174
1998    0.0459

**The following observations can be made for the frequency distribution of the `registration_year` column.**

**We can see that there are some outliers, as indicated by extremely low frequencies for some of the earliest years, such as 1910, 1927, 1929, and 1931. This is because it is very unlikely for a car from these years to still be on the road today, and hence these may represent data entry errors or outliers that need to be further investigated.**

**We can also observe that the frequency of registration gradually increases as we move to more recent years, with the most frequently registered cars being from the 1990s, 2000s, and 2010s. This trend is not unexpected as the number of cars on the road has increased over time, and more people tend to own cars in recent years.** 

**The highest frequency of registration is in the year 2000, with a proportion of 0.0621. We can also see a slight decrease in the frequency of registration in the most recent years, possibly because the data may not be complete for those years.**

In [50]:
# Retrieves top 5 car brands with highest number of used cars for sale listed in 'autos' DataFrame.
top5_brands_listed = autos['brand'].value_counts().head()
print(top5_brands_listed)

volkswagen       9002
bmw              4931
mercedes_benz    4372
opel             4244
audi             3856
Name: brand, dtype: int64


**The brand data shows the number of used cars for sale listed by different car brands. 
The output of the code above shows the top 5 brands with the most listings, with Volkswagen having the highest number of listings at 9002, followed by BMW at 4931, Mercedes-Benz at 4372, Opel at 4244, and Audi at 3856.**

In [79]:
# determine the mean price and mean mileagge for the top 5 brands
top5_brands_average_price = {}
top5_brands_average_mileage = {}

for name in top5_brands_listed.index:
    selected_rows = autos[autos['brand'] == name] 
    average_price = selected_rows['price_usd'].mean()
    average_mileage = selected_rows['odometer_km'].mean()
    top5_brands_average_price[name] = round(average_price)
    top5_brands_average_mileage[name] = round(average_mileage)
    
    
print(top5_brands_average_price, end='')
print("")
print(top5_brands_average_mileage, end='')

{'volkswagen': 5874, 'bmw': 8398, 'mercedes_benz': 8374, 'opel': 3464, 'audi': 9392}
{'volkswagen': 127974, 'bmw': 133000, 'mercedes_benz': 131478, 'opel': 127725, 'audi': 129265}

**We can see that the highest average price is for the `Audi` brand at `$9392`, 
followed by `BMW` and `Mercedes-Benz`, both with similar average prices around `$8374 to $8398`. 
`Volkswagen` has the fourth highest average price at `$5874`, 
while `Opel` has the lowest average price among the top 5 brands at `$3464`.** 


In [83]:
# creating a series for our top 5 brands with mean price and also mean mileage
top5_price_series = pd.Series(top5_brands_average_price)
print("Price Series")
print(top5_price_series)

print("")

print("Mileage Series")
top5_mileage_series = pd.Series(top5_brands_average_mileage)
print(top5_mileage_series)


Price Series
volkswagen       5874
bmw              8398
mercedes_benz    8374
opel             3464
audi             9392
dtype: int64

Mileage Series
volkswagen       127974
bmw              133000
mercedes_benz    131478
opel             127725
audi             129265
dtype: int64


In [87]:
# convert to a single column dataframe 
top5_df = pd.DataFrame(top5_price_series, columns=['average_price'])

# add mileage series to the dataframe 
top5_df['average_mileage'] = top5_mileage_series

# sort by average price
top5_df.sort_values('average_price', ascending=False)

average_price  average_mileage
audi                    9392           129265
bmw                     8398           133000
mercedes_benz           8374           131478
volkswagen              5874           127974
opel                    3464           127725

## Summary

**The data shows the average price and mileage for the top five car brands. 
`Audi` is the most expensive brand on the list, with an average price of `$9392`, while `Opel` is the least expensive, with an average price of `$3464`.** 

**It is interesting to note that despite `Audi` being the most expensive brand, it has the highest average mileage of `129265`, which may suggest that Audi cars are reliable and can last a long time.** 

**On the other hand, `Opel` has the lowest average mileage of `127725`, which may indicate that Opel cars may not be as durable as the other brands.** 

**Overall, the data provides some insights into the relationship between car prices and mileage for the top five car brands listed on the german ebay website.**